cardano-api/src/Cardano/Api/Convenience/Query.hs

```diff
@@ -58,45 +70,100 @@ renderQueryConvenienceError ByronEraNotSupported =
-queryStateForBalancedTx
-  :: CardanoEra era
+queryStateForBalancedTx :: ()
+  => OO.CouldBeF es QueryConvenienceError
+  => OO.CouldBeF es AcquireFailure
+  => OO.CouldBeF es UnsupportedNtcVersionError
+  => OO.CouldBeF es EraMismatch
+  => CardanoEra era
   -> NetworkId
   -> [TxIn]
-  -> IO (Either QueryConvenienceError (UTxO era, ProtocolParameters, EraHistory CardanoMode, SystemStart, Set PoolId))
-queryStateForBalancedTx era networkId allTxIns = runExceptT $ do
-  SocketPath sockPath <- ExceptT $ first SockErr <$> readEnvSocketPath
+  -> ExceptT (OO.Variant es) IO (UTxO era, ProtocolParameters, EraHistory CardanoMode, SystemStart, Set PoolId)
+queryStateForBalancedTx era networkId allTxIns = do
+  SocketPath sockPath <- lift readEnvSocketPath & OO.onLeft (OO.throw . SockErr)
 
   let cModeParams = CardanoModeParams $ EpochSlots 21600
-      localNodeConnInfo = LocalNodeConnectInfo
-                            cModeParams
-                            networkId
-                            sockPath
 
-  qSbe <- except $ getSbe $ cardanoEraStyle era
+  let localNodeConnInfo = LocalNodeConnectInfo cModeParams networkId sockPath
 
-  qeInMode <- toEraInMode era CardanoMode
-    & hoistMaybe (EraConsensusModeMismatch (AnyConsensusMode CardanoMode) (getIsCardanoEraConstraint era $ AnyCardanoEra era))
+  qSbe <- getSbe (cardanoEraStyle era) & OO.hoistEither
 
-  -- Queries
-  let utxoQuery = QueryInEra qeInMode $ QueryInShelleyBasedEra qSbe
-                    $ QueryUTxO (QueryUTxOByTxIn (Set.fromList allTxIns))
-      pparamsQuery = QueryInEra qeInMode
-                        $ QueryInShelleyBasedEra qSbe QueryProtocolParameters
-      eraHistoryQuery = QueryEraHistory CardanoModeIsMultiEra
-      systemStartQuery = QuerySystemStart
-      stakePoolsQuery = QueryInEra qeInMode . QueryInShelleyBasedEra qSbe $ QueryStakePools
+  qeInMode <- toEraInMode era CardanoMode
+    & OO.hoistMaybe (EraConsensusModeMismatch (AnyConsensusMode CardanoMode) (getIsCardanoEraConstraint era $ AnyCardanoEra era))
 
   -- Query execution
-  utxo <- ExceptT $ executeQueryCardanoMode era networkId utxoQuery
-  pparams <- ExceptT $ executeQueryCardanoMode era networkId pparamsQuery
-  eraHistory <- firstExceptT AcqFailure $ ExceptT $ queryNodeLocalState localNodeConnInfo Nothing eraHistoryQuery
-  systemStart <- firstExceptT AcqFailure $ ExceptT $ queryNodeLocalState localNodeConnInfo Nothing systemStartQuery
-  stakePools <- ExceptT $ executeQueryCardanoMode era networkId stakePoolsQuery
+  executeLocalStateQueryExpr_ localNodeConnInfo Nothing $ do
+    utxo <- queryUtxo qeInMode qSbe allTxIns
+    pparams <- queryProtocolParams qeInMode qSbe
+    eraHistory <- queryEraHistory
+    systemStart <- querySystemStart
+    stakePools <- queryStakePools qeInMode qSbe
 
-  return (utxo, pparams, eraHistory, systemStart, stakePools)
+    pure (utxo, pparams, eraHistory, systemStart, stakePools)
```

Comment:
Using `oops`, the query expression is very clean.  Individual queries within the expression can throw any number of errors and without requiring the rewriting of other query handling.

This contrasts to the `ExceptT`/`Either` only solution:  https://github.com/input-output-hk/cardano-node/pull/4825/files#r1083338808